OpenRA.Mods.Common/Projectiles/Railgun.cs

```diff
@@ -0,0 +1,251 @@
+		public readonly int QuantizationCount = 16;
+
+		[Desc("Helix color in (A),R,G,B.")]
+		public readonly Color HelixColor = Color.FromArgb(128, 255, 255, 255);
+		public readonly bool HelixPlayerColor = false;
+
+		[Desc("Impact animation.")]
+		public readonly string HitAnim = null;
+
+		[Desc("Sequence of impact animation to use.")]
+		[SequenceReference("HitAnim")]
+		public readonly string HitAnimSequence = "idle";
+
+		[PaletteReference]
+		public readonly string HitAnimPalette = "effect";
+
+		[Desc("Scan radius for actors damaged by beam. If set to zero (default), it will automatically scale to the largest health shape.",
+			"Only set custom values if you know what you're doing.")]
+		public WDist AreaVictimScanRadius = WDist.Zero;
+
+		[Desc("Scan radius for actors with projectile-blocking trait. If set to zero (default), it will automatically scale",
+			"to the blocker with the largest health shape. Only set custom values if you know what you're doing.")]
+		public WDist BlockerScanRadius = WDist.Zero;
+
+		public IProjectile Create(ProjectileArgs args)
+		{
+			var bc = BeamPlayerColor ? Color.FromArgb(128, args.SourceActor.Owner.Color.RGB) : BeamColor;
+			var hc = HelixPlayerColor ? Color.FromArgb(128, args.SourceActor.Owner.Color.RGB) : HelixColor;
+			return new Railgun(args, this, bc, hc);
+		}
+
+		public void RulesetLoaded(Ruleset rules, WeaponInfo wi)
+		{
+			if (BlockerScanRadius == WDist.Zero)
+				BlockerScanRadius = Util.MinimumRequiredBlockerScanRadius(rules);
+
+			if (AreaVictimScanRadius == WDist.Zero)
+				AreaVictimScanRadius = Util.MinimumRequiredVictimScanRadius(rules);
+		}
+	}
+
+	public class Railgun : IProjectile
+	{
+		readonly ProjectileArgs args;
+		readonly RailgunInfo info;
+		readonly Animation hitanim;
+		public readonly Color BeamColor;
+		public readonly Color HelixColor;
+
+		int ticks = 0;
+		bool animationComplete;
+		WPos target;
+
+		// Lets' save computation: computing these in Railgun instead of RailgunRenderable saves 15 ticks of computation.
+		// Fortunately, railguns, as they are projectiles that is shot only once and that's it, they don't track target.
+		int cycleCnt;
+		WVec sourceToTarget;
+		WVec forwardStep;
+		WVec leftVector;
+		WVec upVector;
+		WAngle angleStep;
+
+		public int CycleCnt { get { return cycleCnt; } }
+		public WVec SourceToTarget { get { return sourceToTarget; } }
+		public WVec ForwardStep { get { return forwardStep; } }
+		public WVec LeftVector { get { return leftVector; } }
+		public WVec UpVector { get { return upVector; } }
+		public WAngle AngleStep { get { return angleStep; } }
+
+		public Railgun(ProjectileArgs args, RailgunInfo info, Color beamColor, Color helixColor)
+		{
+			this.args = args;
+			this.info = info;
+			target = args.PassiveTarget;
+
+			this.BeamColor = beamColor;
+			this.HelixColor = helixColor;
+
+			if (!string.IsNullOrEmpty(info.HitAnim))
+				hitanim = new Animation(args.SourceActor.World, info.HitAnim);
+
+			CalcVectors();
```

Comment:
I was told on IRC that we rather want it as separate function. However, it'd still be nice if you could rename it to e.g. `calculateVectors` (following the naming conventions).